src/App.tsx

```diff
@@ -1,24 +1,283 @@
+    }
+  };
+
+  const handleSingleTodoDelition = async (todoId: number) => {
+    try {
+      setSelectedTodoIds(ids => [...ids, todoId]);
+
+      await deleteTodo(todoId);
+
+      setTodosFromServer(
+        todos => todos.filter(todo => todo.id !== todoId),
+      );
+    } catch {
+      onError(ErrorType.UnableToDelete);
+    } finally {
+      setTimeout(
+        () => {
+          setSelectedTodoIds(
+            ids => ids.filter(id => !(id === todoId)),
+          );
+
+          if (inputRef.current) {
+            inputRef.current.focus();
+          }
+        },
+        300,
+      );
+    }
+  };
+
+  const completedTodoIds = todosFromServer
+    .filter(todo => todo.completed)
+    .map(todo => todo.id);
+
+  const handleCompletedTodosDelition = async () => {
+    const processedIds: number[] = [];
+
+    try {
+      setSelectedTodoIds(ids => [...ids, ...completedTodoIds]);
+
+      completedTodoIds.forEach(async (todoId) => {
+        try {
+          await deleteTodo(todoId);
+          processedIds.push(todoId);
+        } catch {
+          onError(ErrorType.UnableToDelete);
+        }
+      });
+    } catch {
+      onError(ErrorType.UnableToDelete);
+    } finally {
+      setTimeout(
+        () => {
+          setSelectedTodoIds(
+            ids => ids.filter(id => !completedTodoIds.includes(id)),
+          );
+
+          setTodosFromServer(
+            todos => todos.filter(todo => !processedIds.includes(todo.id)),
+          );
+
+          if (inputRef.current) {
+            inputRef.current.focus();
+          }
+        },
+        300,
+      );
+    }
+  };
+
+  const updateTodo = async (updatedTodo: Todo) => {
+    try {
+      setSelectedTodoIds(ids => [...ids, updatedTodo.id]);
+
+      const todoFromResponse = await patchTodo(updatedTodo);
+
+      setTodosFromServer(
+        todos => todos.map(td => {
+          if (td.id === todoFromResponse.id) {
+            return todoFromResponse;
+          }
+
+          return td;
+        }),
+      );
+    } catch {
+      onError(ErrorType.UnableToUpdate);
+      throw new Error(ErrorType.UnableToUpdate);
+    } finally {
+      setSelectedTodoIds(
+        ids => ids.filter(id => !(id === updatedTodo.id)),
+      );
+    }
+  };
+
+  useEffect(() => {
+    setCurrentError(ErrorType.NoError);
+    getTodos(USER_ID)
+      .then(setTodosFromServer)
+      .catch(() => (onError(ErrorType.UnableToLoad)));
+  }, []);
+
+  const filteredTodos = useMemo(
+    () => filterTodos(todosFromServer, selectedFilter),
+    [todosFromServer, selectedFilter],
+  );
+
+  const activeTodos = todosFromServer.filter(todo => !todo.completed);
+  const activeTodoIds = activeTodos.map(todo => todo.id);
+  const activeTodosCount = activeTodos.length;
+
+  const handleToggleAll = () => {
+    if (activeTodosCount) {
+      setSelectedTodoIds(ids => [...ids, ...activeTodoIds]);
+
+      activeTodos.forEach(async (todo) => {
+        try {
+          const updatedTodo = {
```

Comment:
Fix `forEach` with `async` usage in all places

Reply:
I've read the articles you've sent before and those refer to a situation, when calllback function (that is passed to forEach) doesn't have a proper inner structure async (try-catch blocks are missing) hence they can't handle promises. But in my case, the async structure is followed so the promises resolve as expected ![KEKWait (2)](https://github.com/mate-academy/react_todo-app-with-api/assets/146858277/23b682a0-5d2e-4cc4-bcba-ff301590aa20)

Reply:
Not fixed 😢 
@Yevhen-Kolosov those articles generally advise against using `forEach` with asynchronous code